untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,461 @@
+Formally, beta-reduction uses the following rules.
+Given an expression of the form $(\lambda x. t_1)\; t_2$ (i.e. an application of some abstraction to some expression $t_2$):
+
+1. Rename bound variables in $t_1$ and $t_2$ to avoid ambiguity.
+    The new expressions, $t_1'$ and $t_2',$ must be alpha-equivalent to $t_1$ and $t_2,$ respectively.
+2. Replace all the instances of the bound variable $x$ in $t_1'$ with $t_2'.$
+    That is the result of the computation.
+
+Any expression where beta-reduction can be applied, we'll call a redex (**red**ucible **ex**pression).
+
+For example, the expression $(\lambda x. x) y$ is a redex, and one step of beta-reduction transforms it to just $y$.
+
+Per rule 1, we rename bound variables if there might be ambiguity.
+For example, $(\lambda x. \lambda y. x\;y)\;(\lambda x. x\;y)$ is a redex.
+However, the $y$ identifier is ambiguous.
+In the left term, it refers to the bound variable; in the right one, it refers to a free variable.
+To avoid ambiguity, we can rename $y$ in the first expression to, say, $z$: $(\lambda x. \lambda z. x\;z)\;(\lambda x. x\;y).$
+Then we can do the rest of the steps with no fear of confusion: $\lambda z. (\lambda x. x\;y)\;z.$
+
+In the last example, you might wonder if we should also apply beta-reduction inside the abstraction body.
+It brings us to the discussion of evaluation strategies.
+Different evaluation strategies correspond to different computation semantics.
+
+A rule that defines which redexes are reduced in what order is called an evaluation strategy.
+The most straightforward evaluation strategy is called _full beta-reduction._
+It allows us to reduce any redex with no order limitations.
+Full beta-reduction does have some issues, however.
+For instance, you might get different results depending on the reduction order.
+Two common strategies define the order: _normal order strategy_ and _applicative order strategy_.
+The former says that we must first reduce the leftmost _outermost_ redex.
+In other words, we substitute arguments into the abstraction body before beta-reducing them.
+The latter says that we must start with the leftmost _innermost_ redex.
+That is, we reduce the arguments before substitution.
+
+Those are the common _strong_ evaluation strategies.
+We call them strong because they go inside the abstraction bodies.
+The corresponding _weak_ evaluation strategies are called _call-by-name_ and _call-by-value_.
+Call-by-name is a weak version of the normal order strategy.
+Call-by-value is a little stricter and only allows reduction if the argument is a "value" -- that is, in pure lambda calculus, either a variable or an abstraction.
+
+It's interesting to note that the call-by-value strategy corresponds to eager evaluation.
+In eager evaluation, arguments are evaluated before the function call.
+Call-by-name, on the other hand, corresponds to lazy evaluation.
+In lazy evaluation, the arguments potentially are never evaluated at all.
+
```

Comment:
I think this part might require some examples to compare strategies and their pros and cons in the concrete case.

Reply:
It's a little hard to construct examples without first introducing some meaningful terms, which is done below. I could of course use impure calculus (e.g. I could introduce naturals), but doing that for a one-off example seems kind of extraneous.

I'll see what I can do with simple terms, but I'm not very optimistic frankly.

Reply:
Well, I've added _an_ example. It illustrates some usual points, but I can't say I'm very happy with how it turned out. I expect most readers will skip right over.

Reply:
Yeah, I don't think it is really helpful. 

If you use one in the article, I would introduce how that works, tell that there are options, but wouldn't explain this all, it doesn't really help the reader.